odoo/http.py

```diff
@@ -855,866 +1116,459 @@ def not_found(self, description=None):
+                        raise  # bubble up to werkzeug.debug.DebuggedApplication
+
+                    if request.type == 'json':
+                        exc.error_response = self.env['ir.http']._json_handle_error(exc)
+                    else:
+                        exc.error_response = self.env['ir.http']._http_handle_error(exc)
+                    raise
 
-    def render(self):
-        """ Renders the Response's template, returns the result
+    def _serve_ir_http(self):
         """
-        env = request.env(user=self.uid or request.uid or odoo.SUPERUSER_ID)
-        self.qcontext['request'] = request
-        return env["ir.ui.view"]._render_template(self.template, self.qcontext)
-
-    def flatten(self):
-        """ Forces the rendering of the response's template, sets the result
-        as response body and unsets :attr:`.template`
+        Delegate most of the processing to the ir.http model that is
+        extensible by applications.
         """
-        if self.template:
-            self.response.append(self.render())
-            self.template = None
+        ir_http = self.env['ir.http']
 
-class DisableCacheMiddleware(object):
-    def __init__(self, app):
-        self.app = app
+        try:
+            rule, args = ir_http._match(self.httprequest.path)
+        except NotFound:
+            response = ir_http._serve_fallback()
+            if response:
+                self._inject_future_response(response)
+                return response
+            raise
 
-    def __call__(self, environ, start_response):
-        def start_wrapped(status, headers):
-            req = werkzeug.wrappers.Request(environ)
-            root.setup_session(req)
-            if req.session and req.session.debug and not 'wkhtmltopdf' in req.headers.get('User-Agent'):
+        self._session_save = rule.endpoint.routing.get('save_session', True)
+        if not self._session_save:
+            self._set_cookie_session_id(session_id='', dbname=self.db)
 
-                if "assets" in req.session.debug and (".js" in req.base_url or ".css" in req.base_url):
-                    new_headers = [('Cache-Control', 'no-store')]
-                else:
-                    new_headers = [('Cache-Control', 'no-cache')]
+        ir_http._authenticate(rule.endpoint)
+        ir_http._pre_dispatch(rule, args)
+        if self.type == 'json':
+            response = self._json_dispatch(rule.endpoint, args)
+        else:
+            response = self._http_dispatch(rule.endpoint, args)
+        self._inject_future_response(response)
+        return response
 
-                for k, v in headers:
-                    if k.lower() != 'cache-control':
-                        new_headers.append((k, v))
 
-                start_response(status, new_headers)
-            else:
-                start_response(status, headers)
-        return self.app(environ, start_wrapped)
+#----------------------------------------------------------
```

Comment:
There are many competing styles when it comes to do comment titles in our code base:

137 matches:

    # ==== Title ====

123 matches:

    # ====================
    # left-aligned title
    # ====================

83 matches:

    #--------------------
    # left-aligned title
    #--------------------

75 matches:

    # === Title ===

33 matches:

    #=== Title ===

10 matches:

    # ==================== #
    #    centered title    #
    # ==================== #

3 matches:

    # ====================
    # == centered title ==
    # ====================

The style I used throughout this file is a common way to do titles and it was the style used in http.py prior to this refactor: https://github.com/odoo/odoo/blob/1b36b4e8a6b319de10b6c1144c37533dab0e2f36/odoo/http.py#L2-L4

Citing [pep-8](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds):

> Some other good reasons to ignore a particular guideline:
>     To be consistent with surrounding code that also breaks it (maybe for historic reasons) -- although this is also an opportunity to clean up someone else's mess (in true XP style).

Although your linter is crying because there isn't that extra space after the # symbol, it is not a rule that is enforced globally (check style runbot). I know you have your own rules when it comes to website-related modules and I'll make sure the code I change is pylint compliant. But outside (e.g. in core modules) we have our own preferred style that does not strictly comply with pep-8. Still citing pep-8:

> Some other good reasons to ignore a particular guideline:
>     When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.

Those two reasons, (1) consistency and (2) readability counts, will be the reason I also discard other linter errors. Understand me well, I'll not "blindly" discard all of them, you raise this point because it annoys you that have tons of "linter errors" in your editor. Whenever possible I'll comply with your opinion.

About the comment titles, I could use another style, one that is both commonly used and pylint-compliant. May I do so when it doesn't break consistency with the surrounding code. In the specific case of http.py, because I changed most of the file, I guess now is the best time to change it.

Reply:
> The style I used throughout this file is a common way to do titles and it was the style used in http.py prior to this refactor:

Let's not hide behind that one please, especially since 
![image](https://user-images.githubusercontent.com/30048408/142605242-ad866d6d-bf85-44e5-93da-e702efb29dab.png)

> There are many competing styles when it comes to do comment titles in our code base:

Yes, as you said, 116 are wrongly linted and 356 are fine. That's also the purpose of such refactoring, get rid of them, not introduce new ones..

Especially when it is just aesthetic and doesn't bring anything, it is just a comment separator, lint it..

> I know you have your own rules when it comes to website-related modules

Nothing else than regular pep8 linter, we don't use anything else in website-related modules.

Reply:
The thing is, I give much more importance than you do to respect the overall style of the file I'm modifying. "A Foolish Consistency is the Hobgoblin of Little Minds" really does appeal to me. Respecting the existing style within a function, then within a module, then within a project will always be more important than strictly comply with pep8. It is also a matter of not changing code that you didn't write yourself. It really pisses some people when you unilaterally change to style of their files, like they get angry at you and stuff.

> In the specific case of http.py, because I changed most of the file, I guess now is the best time to change it.

Done, also did change the style of the comment titles in the other files I had a lot of changes.